src/test/protocol/KnownFormatToGRPC_test.cpp

```diff
@@ -0,0 +1,924 @@
+            if constexpr (std::is_same_v<KeyType, LedgerEntryType>)
+            {
+                // Fields that gRPC does not include in certain LedgerFormats.
+                //
+                // Some of these may be bugs.
+                static const std::
+                    map<LedgerEntryType, std::vector<SField const*>>
+                        gRPCOmitFields{
+                            {ltACCOUNT_ROOT, {&sfWalletLocator, &sfWalletSize}},
+                            {ltDIR_NODE, {&sfExchangeRate}},
+                            {ltLEDGER_HASHES, {&sfFirstLedgerSequence}},
+                            {ltPAYCHAN, {&sfDestinationNode}},
+                        };
+
+                if (auto const iter = gRPCOmitFields.find(fmtId);
+                    iter != gRPCOmitFields.end())
+                {
+                    std::vector<SField const*> const& omits = iter->second;
+
+                    // Check for fields that gRPC omits from this type.
+                    if (std::find_if(
+                            omits.begin(),
+                            omits.end(),
+                            [&sField](SField const* const omit) {
+                                return *omit == sField;
+                            }) != omits.end())
+                    {
+                        // This is one of the fields that gRPC omits.
+                        continue;
+                    }
+                }
+            }
+
+            // The SFields and gRPC disagree on the names of some fields.
+            // Provide a mapping from SField names to gRPC names for the
+            // known exceptions.
+            //
+            // At least one gRPC spelling is a bug.
+            // clang-format off
+            static const std::map<std::string, std::string> sFieldToGRPC{
+                {"AccountTxnID",      "AccountTransactionID"},
+                {"Fee",               "XRPDropsAmount"},
+                {"PreviousTxnID",     "PreviousTransactionID"},
+                {"PreviousTxnLgrSeq", "PreviousTransactionLedgerSequence"},
+                {"Signature",         "PaymentChannelSignature"},
+                {"SigningPubKey",     "SigningPublicKey"},
+                {"TakerGetsCurrency", "TakerGetsCurreny"},
+                {"TxnSignature",      "TransactionSignature"},
+            };
+            // clang-format on
+
+            std::string gRPCName = sFieldToGRPC.count(sField.getName()) == 1
+                ? sFieldToGRPC.at(sField.getName())
+                : sField.getName();
+
+            sFields.insert({std::move(gRPCName), &sField});
+        }
+        return sFields;
+    }
+
+    // Given a Descriptor for a KnownFormat and a map of the SFields of that
+    // KnownFormat, make sure the fields are aligned.
+    void
+    validateDescriptorAgainstSFields(
+        google::protobuf::Descriptor const* const pbufDescriptor,
+        std::optional<google::protobuf::Descriptor const* const> commonFields,
```

Comment:
Could `commonFields` just be a pointer instead of an optional pointer? It looks like a null pointer works just as well as an un-seated optional in the code.

Reply:
Yup.  Good call.  My initial take was to use a `nullptr`.  But I struggled with whether `nullptr` was acceptable stylistically these days, so I changed it.  The `nullptr` is more straight forward.